README.md

```diff
@@ -0,0 +1,662 @@
+from CDF.
+
+``` python
+>>> def galons_per_minute(val):
+...     return val * 4.40287
+
+>>> specs = {
+...     "flow_in_sm3_per_hour": Tag("FLOW_METER.123"),
+...     "flow_in_galons_per_minute": Tag("FLOW_METER.123").calc(galons_per_minute),
+... }
+
+>>> TagResolver(retrival_function).resolve(specs)
+{
+    "flow_in_sm3_per_hour": 12.3456,
+    "flow_in_galons_per_minute": 54.35604149,
+}
+```
+
+#### Calculations with Multiple Tags
+
+`Tag.call` is a class method that takes a callable and any number of `Tag` instances. When the values are fetched from
+CDF, the callable will be applied (element-wise) with the tag values passed to it as arguments.
+
+``` python
+>>> def closest_to_42(*vals):
+...     """Return whichever value in `vals` is closest to 42"""
+...     deltas = [abs(42 - val) for val in vals]
+...     return vals[deltas.index(min(deltas))]
+
+>>> specs = {
+...     "value_1": Tag("METER_A"),
+...     "value_2": Tag("METER_B"),
+...     "value_3": Tag("METER_C"),
+...     "answer_to_everything": Tag.call(closest_to_42, Tag("METER_A"), Tag("METER_B"), Tag("METER_C")),
+... }
+>>> TagResolver(retrival_function, my_extension).resolve(specs)
+{
+    "value_1": 11,
+    "value_2": 44,
+    "value_3": 57,
+    "answer_to_everything": 44,
+}
+```
+
+#### Referencing Functions by Name
+
+Both `calc` and `call` accept a string instead of a callable for their first argument. In this case, the string
+must match a key in a dict passed to `TagResolver`. This dict contains the actual callables which are then used as
+described above.
+
+This feature can be used to address issues with importing Python modules.
+
+``` python
+>>> my_extension = {
+...     "galons_per_minute": lambda val: return val * 4.40287,
+... }
+
+>>> specs = {
+...     "flow_in_sm3_per_hour": Tag("FLOW_METER.123"),
+...     "flow_in_galons_per_minute": Tag("FLOW_METER.123").calc("galons_per_minute"),
+... }
+
+>>> TagResolver(retrival_function, my_extension).resolve(specs)
+{
+    "flow_in_sm3_per_hour": 12.3456,
+    "flow_in_galons_per_minute": 54.35604149,
+}
+```
+
+``` python
+>>> def closest_to_42(*vals):
+...     """Return whichever value in `vals` is closest to 42"""
+...     deltas = [abs(42 - val) for val in vals]
+...     return vals[deltas.index(min(deltas))]
+
+>>> specs = {
+...     "value_1": Tag("METER_A"),
+...     "value_2": Tag("METER_B"),
+...     "value_3": Tag("METER_C"),
+...     "answer_to_everything": Tag.call("nearest_42", Tag("METER_A"), Tag("METER_B"), Tag("METER_C")),
+... }
+
+>>> TagResolver(retrival_function, {"nearest_42": closest_to_42}).resolve(specs)
+{
+    "value_1": 11,
+    "value_2": 44,
+    "value_3": 57,
+    "answer_to_everything": 44,
+}
+```
+
+### Caching and Combined API Calls
+
+Any call to `TagResolver.resolve` will result in the minimum number of calls to the API to retrieves all needed values.
+
+Each instance of `TagResolver` keeps internal cache and only queries the API for tags that are needed.
```

Comment:
That's a great point! Perhaps we should refuse to execute `.resolve` if the new `specs` dict contains unknown tags. Because if it does, we could get result with `N+1` values whereas other cached tags have only `N` values.
 
Or just don't cache nothing never and simply query CDF? 🤔

Reply:
(cannot to element-wise operations between series with N and N+1 values, is the issue)

Reply:
Hmm... not sure what makes most sense - complexity vs. performance.. 🤔 

(Another option could be to clear the cache and re-execute the cache if you get into the "N and N+1"-state)

Reply:
Yeah, complexity is my primary concern here, don't want to over-engineer it, and it's getting close...

If we want to check whether we are in this "N & N+1" situation, we have to either make a separate query OR we can assume that there is no N+1 and act only if we see it in the data.

Option 1 is not really a good solution, it just narrows the window for new points to appear, but it doesn't eliminate it (there is still some non-zero time between the check and the query).

Option 2 is kind of wasteful, because if we detect the N+1 case, we basically have to throw away the data and repeat the query with more tags. Also, complexity is piling up...

We could just drop any caching, but that has problems too. Specifically, I would like to still be able to do things like this:

``` python
resolver = TagResolver(blah, blah)
values_specs = {
    "foo": Tag("FOO"),
    "bar": Tag("BAR"),
}
percent_specs = {
    "foo": Tag("FOO") * 100 / (Tag("FOO") + Tag("BAR")),
    "bar": Tag("BAR") * 100 / (Tag("FOO") + Tag("BAR")),
}
values = resolver.resolve(values_specs)
percentages = resolver.resolve(percent_specs)
```

It is important to not query CDF 2 times here (like if we just dropped all caching in `TagResolver`), because if we do 2 queries, the values in `percentages` might mot match those in `values` dict any more.

So I'm inclined to go with the solution where an instance of `TagResolver` can only perform a single query to CDF and that's it. With that solution, introducing new tags in later calls to `resolve` would not work, for example:

``` python
resolver = TagResolver(blah, blah)
values_specs = {
    "foo": Tag("FOO"),
    "bar": Tag("BAR"),
}
percent_specs = {
    "foo": Tag("FOO") * 100 / Tag("BAZZZZZZ"),
    "bar": Tag("BAR") * 100 / Tag("BAZZZZZZ"),
}
values = resolver.resolve(values_specs)  # ok
percentages = resolver.resolve(percent_specs)  # 💣 💥 
```
`💣 💥 ` meaning we raise something like UnresolvedTagError("Tag BAZZZZZ has not been resolved. Any subsequent calls to `resolve()` must not introduce new tags. Instead create a new instance of `TagResolver`.")

Not sure... thoughts very much welcome!

Reply:
(Just a note) this would however work: 

```py
percentages = resolver.resolve(percent_specs) # ok
values = resolver.resolve(values_specs) # ok - no new tags
```

Reply:
Yeap